### DISCUSSION TITLE
How to use new mask feature?

### DISCUSSION THREAD

**Discussion by MoreColors123** [Acly/krita-ai-diffusion#557] [Q&A]
I tried but failed, how do I use this exactly? Could someone give me a hint?
![grafik](https://github.com/Acly/krita-ai-diffusion/assets/131474336/b33bc00b-7745-4685-8e56-a1d61cccf768)

**Comment by MoreColors123** [comment]
Did anyone already use this feature? I'm still not getting it.

**Comment by Acly** [comment]
Key things:
* Use Selection Masks to create your mask. You can edit them with soft brush, or gradient fill, etc.
* Use 90% Refine, or consider 100% Custom inpaint and disable seamless if you want more "drastic" changes in the image


https://github.com/Acly/krita-ai-diffusion/assets/6485914/e46c35d2-bc7e-4159-a252-4a775bd37717

**Reply by SyamsQ** [reply]
I use retangular selection, but red area doesn't appears. How to do it.?

**Comment by Muskelmagier** [comment]
while it probably is a nice feature but what it comes down to is that I have to manually soften the edges of every inpaint I do otherwise I have visible edges in my pictures.

**Reply by Acly** [reply]
Please post pictures/workflows that can ideally reproduced. I have a decently sized benchmark but not seen any issues. Mostly transitions improved slightly because it's now blending denoising strength, not alpha (which is more of a hack).

**Reply by Muskelmagier** [reply]
Raw
![raw](https://github.com/Acly/krita-ai-diffusion/assets/59380143/f585d800-a2f9-4bf8-a11a-fca3f40258aa)
after inpaint strength: 40%
![inpainted](https://github.com/Acly/krita-ai-diffusion/assets/59380143/4983612f-9708-45b8-be20-f15b2d5760ed)

raw, 
![raw1](https://github.com/Acly/krita-ai-diffusion/assets/59380143/b98021b3-7e2e-4133-8002-e430d9c1af6e)
after inpaint, strength: 40%
![inpainted1](https://github.com/Acly/krita-ai-diffusion/assets/59380143/674ae365-d356-437b-8e19-bedd41888cc3)

a few examples I created in 3 minutes. it's not always there because it heavily depends on what kind of region you inpaint in and how many different colors are present, but when you have a big complex piece it gets increasingly more visible due to the slight color shift rims everywhere.

**Reply by Acly** [reply]
Thanks. I'm guessing it's an issue in those situations where for some reason the inpainting cannot match the colors of the context at all. Even a low denoising strength is enough to allow subtle color shift, which become more obvious when they are over a large region.

If you open [comfy_workflow.py](https://github.com/Acly/krita-ai-diffusion/blob/style-fixes/ai_diffusion/comfy_workflow.py#L493) and find the line 
```
def denoise_to_compositing_mask(self, mask: Output, offset=0.15, threshold=0.25):
```
you can tweak the values which determine the ratio between "soft inpainting" (differential diffusion) vs. alpha blending.

If you set `offset=0` and `threshold=1` you should get pure alpha blending (original behavior). But soft inpainting would no longer work.

It would be interesting to get values that still provide good results in all cases. Try something like `offset=0.1` and `threshold=0.35` etc.

(And yes we can make a slider, but that's always more work, and good defaults are still important regardless)
